### ISSUE TITLE
UnboundLocalError when module is shadowed by a comprehension and then accessed by another comprehension

### ISSUE THREAD

**Issue by jmymay** [python/cpython#118513]
# Bug report

### Bug description:

Seems to be a regression in 3.12 when compared to 3.11, likely caused by [PEP 709: Comprehension inlining](https://docs.python.org/3.12/whatsnew/3.12.html#pep-709-comprehension-inlining), but I haven't verified this. Example below. (Noticed while running [this code](https://github.com/Vocab-Apps/anki-hyper-tts/blob/5a1637c/service.py#L74).)

```python
import sys
def f():
    _ = [sys for sys in []]  # can be any module other than sys
    [sys.stdout.write('success\n') for _ in [1]]
f()
```

```
$ python3.11 main.py  # 3.11.9
success
```

```
$ python3.12 main.py  # 3.12.3
Traceback (most recent call last):
  File "/home/jkozera/imp/main.py", line 5, in <module>
    f()
  File "/home/jkozera/imp/main.py", line 4, in f
    [sys.stdout.write('success\n') for _ in [1]]
     ^^^
UnboundLocalError: cannot access local variable 'sys' where it is not associated with a value
```

```
$ python3.13 main.py  # 3.13.0a6+ (heads/main:2770d5c, May  2 2024, 18:19:03)
Traceback (most recent call last):
  File "/home/jkozera/imp/main.py", line 5, in <module>
    f()
    ~^^
  File "/home/jkozera/imp/main.py", line 4, in f
    [sys.stdout.write('success\n') for _ in [1]]
     ^^^
UnboundLocalError: cannot access local variable 'sys' where it is not associated with a value
```

### CPython versions tested on:

3.11, 3.12, CPython main branch

### Operating systems tested on:

Linux

### Linked PRs
* gh-118526
* gh-118548

**Comment by Eclips4** [comment]
cc @carljm

**Comment by JelleZijlstra** [comment]
Confirmed without a module too:
```
x = 1
def f():
    [x for x in [1]]
    return [x for _ in [1]]
print(f())
```

```
Traceback (most recent call last):
  File "/Users/jelle/py/cpython/globs.py", line 5, in <module>
    print(f())
          ~^^
  File "/Users/jelle/py/cpython/globs.py", line 4, in f
    return [x for _ in [1]]
            ^
UnboundLocalError: cannot access local variable 'x' where it is not associated with a value
```

Disassembly shows we're using `LOAD_FAST_CHECK` for the second comprehension when we should be using `LOAD_GLOBAL`.

**Comment by Eclips4** [comment]
> Confirmed without a module too:
> 
> ```
> x = 1
> def f():
>     [x for x in [1]]
>     return [x for _ in [1]]
> print(f())
> ```
> 
> ```
> Traceback (most recent call last):
>   File "/Users/jelle/py/cpython/globs.py", line 5, in <module>
>     print(f())
>           ~^^
>   File "/Users/jelle/py/cpython/globs.py", line 4, in f
>     return [x for _ in [1]]
>             ^
> UnboundLocalError: cannot access local variable 'x' where it is not associated with a value
> ```
> 
> Disassembly shows we're using `LOAD_FAST_CHECK` for the second comprehension when we should be using `LOAD_GLOBAL`.

FYI, the assignment ``x = 1`` here's unnecessary.

**Comment by carljm** [comment]
Yes, either way it should compile to `LOAD_GLOBAL`; without the global assignment it should result in `NameError`, not `UnboundLocalError`.

**Comment by carljm** [comment]
Fix in #118526
